File: day4/day4.go
package main

import "fmt"

func main() {
	const start = 246540
	const end = 787419

	totalOne := 0
	totalTwo := 0
	var digits []uint8

	for n := start; n <= end; n++ {
		digits = getDigits(n)

		if meetsCriteriaOne(digits) {
			totalOne++
		}

		if meetsCriteriaTwo(digits) {
			totalTwo++
		}

	}

	fmt.Println(totalOne)
	fmt.Println(totalTwo)
}

func meetsCriteriaOne(digits []uint8) bool {
	foundAdjacent := false

	for i := 1; i < len(digits); i++ {
		// digits are in reverse order. So, non-decreasing means
		// every digits at idx i-1 should be >= digit at idx i
		if digits[i-1] < digits[i] {
			return false
		}

		if digits[i] == digits[i-1] {
			foundAdjacent = true
		}
	}

	return foundAdjacent
}

func meetsCriteriaTwo(digits []uint8) bool {
	adjacent := make(map[uint8]uint8)

	for i := 1; i < len(digits); i++ {
		// digits are in reverse order. So, non-decreasing means
		// every digits at idx i-1 should be >= digit at idx i
		if digits[i-1] < digits[i] {
			return false
		}

		if digits[i] == digits[i-1] {
			tmp := digits[i]*10 + digits[i]
			adjacent[tmp]++
		}
	}

	for _, freq := range adjacent {
		if freq == 1 {
			return true
		}
	}

	return false
}

func getDigits(num int) (digits []uint8) {
	for num > 0 {
		digits = append(digits, uint8(num%10))
		num = num / 10
	}

	// No need to correct the digit order, as it doesn't matter
	//length := len(digits)
	//for i := 0; i < ((length + 1) >> 1); i++ {
	//tmp := digits[i]
	//digits[i] = digits[length - i - 1]
	//digits[length - i - 1] = tmp
	//}

	return
}

File: day4/day4.py
from typing import List

start = 246540
end = 787419


def get_digits(n: int) -> List[int]:
    digits = []

    while n > 0:
        digits.append(n % 10)
        n //= 10

    digits.reverse()
    return digits


def isNotPartOfGroup(idx: int, l: int, digits: List[int]) -> bool:
    d = digits[idx]
    if idx == 1:
        return d != digits[idx] + 1
    if idx == l:
        return d != digits[idx - 2]

    return d != digits[idx - 2] and d != digits[idx + 1]


def meets_criteria(num: int) -> bool:
    digits = get_digits(num)
    adjacent_seen = {}
    for i in range(1, len(digits)):
        # should be non-decreasing
        if digits[i] < digits[i - 1]:
            return False

        if digits[i] == digits[i - 1]:
            k = (digits[i], digits[i])
            adjacent_seen[k] = adjacent_seen.get(k, 0) + 1

    return any(x == 1 for x in adjacent_seen.values())


print(meets_criteria(112233))
print(meets_criteria(123444))
print(meets_criteria(11111122))

total = 0
for num in range(start, end + 1):
    if meets_criteria(num):
        total += 1

print(total)
